op-batcher/batcher/espresso.go

```diff
@@ -39,140 +35,285 @@ const (
+
+	l.channelMgrMutex.Lock()
+	defer l.channelMgrMutex.Unlock()
+	syncActions, outOfSync := computeSyncActions(*newSyncStatus, l.prevCurrentL1, l.channelMgr.blocks, l.channelMgr.channelQueue, l.Log, l.Config.PreferLocalSafeL2)
+	if outOfSync {
+		l.Log.Warn("Sequencer is out of sync, retrying next tick.")
+		return
 	}
+	l.prevCurrentL1 = newSyncStatus.CurrentL1
+	if syncActions.clearState != nil || shouldClearState {
+		l.channelMgr.Clear(*syncActions.clearState)
+	} else {
+		l.channelMgr.PruneSafeBlocks(syncActions.blocksToPrune)
+		l.channelMgr.PruneChannels(syncActions.channelsToPrune)
+	}
+}
 
-	timer := time.NewTimer(transactionFetchTimeout)
-	defer timer.Stop()
+// Periodically refreshes the sync status and polls Espresso streamer for new batches
+func (l *BatchSubmitter) espressoBatchLoadingLoop(ctx context.Context, wg *sync.WaitGroup, publishSignal chan struct{}) {
+	l.Log.Info("Starting EspressoBatchLoadingLoop")
 
-	ticker := time.NewTicker(transactionFetchInterval)
+	defer wg.Done()
+	ticker := time.NewTicker(l.Config.PollInterval)
 	defer ticker.Stop()
 
-	var txQueryData espressoCommon.TransactionQueryData
-Loop:
+	streamer := espresso.EspressoStreamer{
+		BatcherAddress: l.BatcherAddress,
+		Namespace:      l.RollupConfig.L2ChainID.Uint64(),
+
+		L1Client:            l.L1Client,
+		EspressoClient:      l.Espresso,
+		EspressoLightClient: l.EspressoLightClient,
+		Log:                 l.Log,
+
+		BatchPos: 1,
+	}
+
 	for {
 		select {
 		case <-ticker.C:
-			txQueryData, err = l.Espresso.FetchTransactionByHash(l.shutdownCtx, txHash)
-			if err == nil {
-				break Loop
+			newSyncStatus, err := l.getSyncStatus(ctx)
+			if err != nil {
+				l.Log.Error("failed to refresh sync status", "err", err)
+				continue
 			}
-			l.Log.Warn("Retry fetching transaction by hash", "txHash", txHash, "error", err)
-		case <-timer.C:
-			l.Log.Error("Failed to fetch transaction by hash after multiple attempts", "txHash", txHash)
-			l.recordFailedDARequest(txdata.ID(), err)
-			return nil, fmt.Errorf("failed to fetch transaction by hash: %w", err)
+
+			l.espressoSyncAndRefresh(ctx, newSyncStatus, &streamer)
+
+			err = streamer.Update(ctx)
+			if err != nil {
+				l.Log.Error("failed to update Espresso streamer", "err", err)
+				continue
+			}
+
+			var batch *espresso.EspressoBatch
+			for {
+				batch = streamer.Next(ctx)
+				if batch == nil {
+					break
+				}
+
+				block, err := espresso.BatchToIncompleteBlock(l.RollupConfig, batch)
+				if err != nil {
+					l.Log.Error("failed to convert singular batch to block", "err", err)
+					continue
+				}
+
+				l.Log.Debug("Received block from Espresso", "blockNr", block.NumberU64(), "blockHash", block.Hash(), "parentHash", block.ParentHash())
+
+				l.channelMgrMutex.Lock()
+				err = l.channelMgr.AddL2Block(block)
+				l.channelMgrMutex.Unlock()
+
+				if err != nil {
+					l.Log.Error("failed to add L2 block to channel manager", "err", err)
+					l.clearState(ctx)
+				}
```

Comment:
If we are unable to add the block I think we need to retry or something, as again we don't want to skip this block due to errors that wouldn't be replicated in the caff node

Reply:
The only error state is if the block doesn't extend the tip we already have in state, so we clear state and start again from known safe L2. Should've been resetting the streamer here as well though.